AppCenterDistribute/AppCenterDistribute/MSDistribute.m

```diff
@@ -589,20 +590,33 @@ - (void)openURLInAuthenticationSessionWith:(NSURL *)url API_AVAILABLE(ios(11)) {
     if (callbackUrl) {
       [strongSelf openURL:callbackUrl];
     }
+    [[MSUtility sharedApp] endBackgroundTask:backgroundAuthSessionTask];
   };
   SFAuthenticationSession *session = [[SFAuthenticationSession alloc] initWithURL:url
                                                                 callbackURLScheme:callbackUrlScheme
                                                                 completionHandler:authCompletionBlock];
 
+  if (self.authenticationSession != nil) {
+
+    // Calling 'start' on an existing session crashes the application - clear session.
+    [self clearAuthenticationSession];
+  }
+  
   // Retain the session.
   self.authenticationSession = session;
-  @try {
-    BOOL success = [session start];
-    if (success) {
-      MSLogDebug([MSDistribute logTag], @"Authentication session started, showing confirmation dialog.");
-    }
-  } @catch (NSException *exception) {
-    MSLogError([MSDistribute logTag], @"Failed to start authentication session: %@", exception.reason);
+
+  /* Request additional background execution time for authorization. If we authorize using third-party services (MS Authenticator)
+   * then switching to another application will kill the current session. This line fixes this problem.
+   */
+  backgroundAuthSessionTask = [[MSUtility sharedApp] beginBackgroundTaskWithName:@"Safari authentication"
+                                                               expirationHandler:^{
+                                                                 [[MSUtility sharedApp] endBackgroundTask:backgroundAuthSessionTask];
```

Comment:
We should cancel the session here instead of calling `endBackgroundTask`. Then, `endBackgroundTask` and `strongSelf.authenticationSession = nil` will happen inside the completion handler of the session. Although I suppose it would need to be called from the main thread, would that cause a problem at this point of execution?
Also, this should probably be the only place where we need to cancel the session.

Reply:
Yes, I think you're right, I will correct this point.